common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -439,49 +459,69 @@ LoopDispatch:
 			continue LoopDispatch
 		}
 
-		now := s.timeSource.Now()
-		reservation := rateLimiter.Reserve(
-			now,
-			channel.rateLimitRequest,
-		)
-		if reservation.DelayFrom(now) != 0 {
-			reservation.CancelAt(now)
-			channel.throttled = true
-			numThrottled++
-			if numThrottled == len(iwrrChannels.channels) {
-				// all channels throttled
-				break LoopDispatch
-			}
+		if len(channel.Chan()) == 0 {
 			continue LoopDispatch
 		}
+
+		if !rateLimiter.Allow(s.timeSource.Now(), channel.rateLimitRequest) {
+			s.metricsHandler.Counter(metrics.TaskSchedulerThrottled.GetMetricName()).Record(1, channel.metricsTags...)
+
+			if rateLimiterConfig.enableThrottle {
+				channel.throttled = true
+				numThrottled++
+				if numThrottled == len(iwrrChannels.channels) {
+					// all channels throttled
+					break LoopDispatch
+				}
+				continue LoopDispatch
+			}
+
+			// throttled, but in shadow mode, do not actually throttle task dispatching
+		}
+
 		select {
 		case task := <-channel.Chan():
 			s.fifoScheduler.Submit(task)
-			numTasks++
-			channel.moreTasks = true
+			taskDispatched++
 		default:
-			reservation.CancelAt(now)
-			channel.moreTasks = false
 			continue LoopDispatch
 		}
 	}
-	atomic.AddInt64(&s.numInflightTask, -numTasks)
+	atomic.AddInt64(&s.numInflightTask, -taskDispatched)
```

Comment:
Consider making this value an Int64 instead to prevent some rare bugs, as noted in the docs for the atomic package: https://pkg.go.dev/sync/atomic#AddInt64

Reply:
Both `taskDispatched` and `numInflightTask` are already defined as int64 I believe.